src/viewers/VivViewer.js

```diff
@@ -0,0 +1,164 @@
+    navigator.userAgent.indexOf('CriOS') === -1 &&
+    navigator.userAgent.indexOf('FxiOS') === -1
+  );
+}
+
+/**
+ * This component handles rendering the various views within the DeckGL contenxt.
+ * @param {Array} layerProps The props for the layers in each view.
+ * @param {Array} randomize Wgether or not to randomize which view goes first (for dynamic rendering).
+ * @param {VivView} views The various VivViews to render.
+ * */
+export default class VivViewer extends PureComponent {
+  constructor(props) {
+    super(props);
+    this.state = {
+      viewStates: {}
+    };
+    const { viewStates } = this.state;
+    const { views } = this.props;
+    views.forEach(view => {
+      viewStates[view.id] = view.filterViewState({
+        viewState: view.initialViewState
+      });
+    });
+    this._onViewStateChange = this._onViewStateChange.bind(this);
+    this.layerFilter = this.layerFilter.bind(this);
+  }
+
+  /**
+   * This prevents only the `draw` call of a layer from firing,
+   * but not other layer lifecycle methods.  Nonetheless, it is
+   * still useful.
+   * @param {Layer} layer The layer being updated.
+   * @param {Viewport} viewport The viewport being updated.
+   * @returns {boolean} Whether or not this layer should be drawn in this viewport.
+   */
+  // eslint-disable-next-line class-methods-use-this
+  layerFilter({ layer, viewport }) {
+    return layer.id.includes(getVivId(viewport.id));
+  }
+
+  /**
+   * This updates the viewState as a callback to the viewport changing in DeckGL
+   * (hence the need for storing viewState in state).
+   */
+  _onViewStateChange({ viewId, viewState, oldViewState }) {
+    // Save the view state and trigger rerender.
+    const { views } = this.props;
+    this.setState(prevState => {
+      const viewStates = {};
+      views.forEach(view => {
+        const currentViewState = prevState.viewStates[view.id];
+        viewStates[view.id] = view.filterViewState({
+          viewState: { ...viewState, id: viewId },
+          oldViewState,
+          currentViewState
+        });
+      });
+      return { viewStates };
+    });
+  }
+
+  /**
+   * This updates the viewStates' height and width with the newest height and
+   * width on any call where the viewStates changes (i.e resize events),
+   * using the previous state (falling back on the view's initial state) for target x and y, zoom level etc.
+   */
+  componentDidUpdate(prevProps) {
+    const { views } = this.props;
+    // Update internal viewState on view changes as well as height and width changes.
+    // Maybe we should add x/y too?
+    if (
+      views.some(
+        (view, i) =>
+          prevProps.views[i] !== view ||
+          view.initialViewState.height !==
+            prevProps.views[i].initialViewState.height ||
+          view.initialViewState.width !==
+            prevProps.views[i].initialViewState.width
+      )
+    ) {
+      // eslint-disable-next-line react/no-did-update-set-state
```

Comment:
I got `getDerivedStateFromProps` working (I had tried `shouldComponentUpdate` but it wasn't right for this) - thanks for the suggestion!  I found that the functional component works but was somewhat slower updating the `deck.gl` stuff for resizing, and `React.memo` didn't really help because so many different things can change (sliders, colors, colormaps etc.).  None of the `deck.gl` demos use them and I wonder if there is a reason....

In any case, I'm going to file an issue - this is a nice first pass, but I think someone should investigate relative sizing the `DeckGL` component with our use cases.  I know it is in theory possible with `deck.gl` but we do a lot of custom stuff that might make it infeasible/nasty.  I wonder if this change to how sizing is performed would help the functional component performance.

Reply:
Great! Generally if I'm writing things in React, it's my preference now to use functional components, but for more complex state I've run into cases where it's easier to use a class and specific lifecycle methods. The concern here was not using the correct lifecycle method and ignoring eslint rules. We certainly can go over this again and refactor, but this seems like a totally appropriate solution. Thanks.

Reply:
FWIW, in the end I don't think the trade-offs between class and functional components are usually performance based. Spending enough time, there is usually a way to achieve similar effects. I prefer functional simply because they are easier to read (IMO) and generally result in less code (so easier to maintain).